assets/ca/ca.querytranslator.js

```diff
@@ -0,0 +1,387 @@
+			},
+			{
+				regex: /^\[\s*(.*)\s+TO\s+(.*)\s*]$/,
+				operator: 'between',
+				negatedOperator: 'not_between'
+			},
+			{
+				regex: /^([^\*].*[^\\])\*$/,
+				operator: 'begins_with',
+				negatedOperator: 'not_begins_with'
+			},
+			{
+				regex: /^\*(.*[^\\])\*$/,
+				operator: 'contains',
+				negatedOperator: 'not_contains'
+			},
+			{
+				regex: /^\*(.*(?:\\\*|[^\*]))$/,
+				operator: 'ends_with',
+				negatedOperator: 'not_ends_with'
+			},
+			{
+				regex: /^\*$/,
+				operator: 'is_empty',
+				negatedOperator: 'is_not_empty'
+			}
+		];
+
+	/**
+	 * Escape the user-entered field value.
+	 * @param {String} value
+	 * @returns {String}
+     */
+	escapeValue = function (value) {
+		return value.replace(/([\-\+&\|!\(\)\{}\[\]\^"~\*\?:\\])/, '\\$1');
+	};
+
+	/**
+	 * Convert a set of rules from the jQuery query builder into a CA search query.  Performs the inverse operation to
+	 * `convertSearchQueryToQueryBuilderRules`.
+	 * @param {Object} ruleSet
+	 * @returns {String}
+	 */
+	caUI.convertQueryBuilderRuleSetToSearchQuery = function (ruleSet) {
+		if (ruleSet.condition && ruleSet.rules) {
+			return '(' + $.map(ruleSet.rules, caUI.convertQueryBuilderRuleSetToSearchQuery).join(' ' + ruleSet.condition + ' ') + ')';
+		}
+		if (ruleSet.operator && ruleSet.field) {
+			// Escape value to allow special characters
+			var negation = ruleSet.operator.match(/not_/),
+				prefix = ruleSet.field + (negation ? ':-' : ':');
+			switch (negation ? ruleSet.operator.replace('not_', '') : ruleSet.operator) {
+				case 'equal':
+					return prefix + '"' + escapeValue(ruleSet.value) + '"';
+				case 'in':
+					return prefix + '(' + escapeValue(ruleSet.value) + ')';
+				case 'between':
+					return prefix + '[' + escapeValue(ruleSet.value[0]) + ' TO ' + escapeValue(ruleSet.value[1]) + ']';
+				case 'begins_with':
+					return prefix + '"' + escapeValue(ruleSet.value) + '*"';
+				case 'contains':
+					return prefix + '"*' + escapeValue(ruleSet.value) + '*"';
+				case 'ends_with':
+					return prefix + '"*' + escapeValue(ruleSet.value) + '"';
+				case 'is_empty':
+				case 'is_null':
+					return prefix + '*';
+			}
+			return ruleSet.field + ':' + ruleSet.value;
+		}
+		return '';
+	};
+
+	/**
+	 * Retrieve the list of tokens from the given query string.
+	 * @param {String} query
+	 * @return {Array}
+     */
+	getTokenList = function (query) {
+		var token,
+			tokens = [],
+			queryArray = query.trim().split('');
+		while (token = shiftToken(queryArray)) {
+			tokens.push(token);
+		}
+		while (tokens.length >= 2 && tokens[0].type === TOKEN_LPAREN && tokens[tokens.length - 1].type === TOKEN_RPAREN) {
+			tokens = tokens.slice(1, tokens.length - 1);
+		}
+		return tokens;
+	};
+
+	/**
+	 * Retrieve the next token from the `queryArray`, which is destructively processed.
+	 * @param {Array} queryArray
+	 * @returns {Object|undefined}
+	 */
+	shiftToken = function (queryArray) {
+		var character, token,
+			quoted = false,
```

Comment:
This is what happens when you use tabs for indentation.  I might change this to a single-line declaration and then assign values on the following lines.  (Same for other functions).